cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -710,12 +732,42 @@ public int deleteAllIndexes(int limit) {
     return count;
   }
 
-  // TODO: CDAP-5663 The entire logic of mapping between Indexers and keys should be made internal to MetadataDataset
-  private Indexer getIndexerForKey(String key) {
-    if ("schema".equals(key)) {
-      return new SchemaIndexer();
+  /**
+   * Returns all the {@link Indexer indexers} that apply to a specified metadata key.
+   *
+   * @param key the key for which {@link Indexer indexers} are required
+   */
+  private Set<Indexer> getIndexersForKey(String key) {
+    // only DefaultIndexer for user scope
+    if (MetadataScope.USER == scope) {
+      return Collections.<Indexer>singleton(new DefaultValueIndexer());
+    }
+    Set<Indexer> indexers = new HashSet<>();
+    switch (key) {
+      case AbstractSystemMetadataWriter.SCHEMA_KEY:
+        indexers.add(new SchemaIndexer());
+        break;
+      case AbstractSystemMetadataWriter.ENTITY_NAME_KEY:
+        // used for listing entities sorted in ascending order of name
+        indexers.add(new ValueOnlyIndexer());
```

Comment:
rather than building this set every time, it may be better to create a Map<String, Set<Indexer>>, then just perform a lookup. 

I think it would also be better to return both the indexers and the index column for that indexer here (or add a method to Indexer to return the column), since you have all the information required here. That way we could get rid of that getIndexColumn method and its uses of instanceof.